agentendpoint/config_task.go

```diff
@@ -0,0 +1,459 @@
+			ConfigAssignment: a.GetConfigAssignment(),
+			Result: &agentendpointpb.ApplyConfigTaskOutput_ConfigAssignmentResult_OsPolicyResults{
+				OsPolicyResults: &agentendpointpb.ApplyConfigTaskOutput_OSPolicyResults{
+					Results: make([]*agentendpointpb.ApplyConfigTaskOutput_OSPolicyResult, len(a.GetPolicies())),
+				},
+			},
+		}
+		c.results.GetResults()[i] = aResult
+		for i, p := range a.GetPolicies() {
+			pResult := &agentendpointpb.ApplyConfigTaskOutput_OSPolicyResult{
+				Id: p.GetId(),
+				Result: &agentendpointpb.ApplyConfigTaskOutput_OSPolicyResult_ResourceResults{
+					ResourceResults: &agentendpointpb.ApplyConfigTaskOutput_ResourceResults{
+						Results: make([]*agentendpointpb.ApplyConfigTaskOutput_ResourceResult, len(p.GetResources())),
+					},
+				},
+			}
+			aResult.GetOsPolicyResults().GetResults()[i] = pResult
+			for i, r := range p.GetResources() {
+				pResult.GetResourceResults().GetResults()[i] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult{
+					Id:             r.GetId(),
+					ExecutionSteps: make([]*agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep, numExecutionSteps),
+				}
+			}
+		}
+	}
+}
+
+func (c *configTask) run(ctx context.Context) error {
+	clog.Infof(ctx, "Beginning apply config task.")
+	c.StartedAt = time.Now()
+	rcsErrMsg := "Error reporting continuing state"
+	if err := c.reportContinuingState(ctx, agentendpointpb.ApplyConfigTaskProgress_STARTED); err != nil {
+		return c.handleErrorState(ctx, rcsErrMsg, err)
+	}
+
+	if len(c.Task.GetConfig().GetConfigAssignments()) == 0 {
+		clog.Infof(ctx, "No config assignments to apply.")
+		return c.reportCompletedState(ctx, "", agentendpointpb.ApplyConfigTaskOutput_SUCCEEDED)
+	}
+
+	// We need to generate base results first thing, each execution step
+	// just adds on.
+	c.generateBaseResults()
+
+	if err := c.reportContinuingState(ctx, agentendpointpb.ApplyConfigTaskProgress_VALIDATING); err != nil {
+		return c.handleErrorState(ctx, rcsErrMsg, err)
+	}
+	c.validation(ctx)
+
+	if err := c.reportContinuingState(ctx, agentendpointpb.ApplyConfigTaskProgress_CHECKING_DESIRED_STATE); err != nil {
+		return c.handleErrorState(ctx, rcsErrMsg, err)
+	}
+	c.checkState(ctx)
+
+	if err := c.reportContinuingState(ctx, agentendpointpb.ApplyConfigTaskProgress_ENFORCING_DESIRED_STATE); err != nil {
+		return c.handleErrorState(ctx, rcsErrMsg, err)
+	}
+	c.enforceState(ctx)
+
+	if c.postCheckRequired {
+		if err := c.reportContinuingState(ctx, agentendpointpb.ApplyConfigTaskProgress_CHECKING_DESIRED_STATE_POST_ENFORCEMENT); err != nil {
+			return c.handleErrorState(ctx, rcsErrMsg, err)
+		}
+		c.postCheckState(ctx)
+	}
+
+	return c.reportCompletedState(ctx, "", agentendpointpb.ApplyConfigTaskOutput_SUCCEEDED)
+}
+
+// RunApplyConfig runs an apply config task.
+func (c *Client) RunApplyConfig(ctx context.Context, task *agentendpointpb.Task) error {
+	ctx = clog.WithLabels(ctx, task.GetServiceLabels())
```

Comment:
add task id to context?

Reply:
Thats what this should do, any labels the service sends down.